contracts/metatx/MinimalForwarder.sol

```diff
@@ -23,82 +29,202 @@ contract MinimalForwarder is EIP712 {
+        );
 
-    mapping(address => uint256) private _nonces;
+    /**
+     * @dev Emitted when a `ForwardRequest` is executed.
+     */
+    event ExecutedForwardRequest(address indexed signer, uint256 nonce, bool success, bytes returndata);
 
     /**
      * @dev The request `from` doesn't match with the recovered `signer`.
      */
     error MinimalForwarderInvalidSigner(address signer, address from);
 
     /**
-     * @dev The request nonce doesn't match with the `current` nonce for the request signer.
+     * @dev The requested `value` doesn't match with the available `msgValue`, leaving ETH stuck in the contract.
      */
-    error MinimalForwarderInvalidNonce(address signer, uint256 current);
+    error MinimalForwarderMismatchedValue(uint256 value, uint256 msgValue);
 
-    constructor() EIP712("MinimalForwarder", "0.0.1") {}
+    /**
+     * @dev The list of requests length doesn't match with the list of signatures length.
+     */
+    error MinimalForwarderInvalidBatchLength(uint256 requestsLength, uint256 signaturesLength);
 
-    function getNonce(address from) public view returns (uint256) {
-        return _nonces[from];
-    }
+    /**
+     * @dev The request `deadline` has expired.
+     */
+    error MinimalForwarderExpiredRequest(uint256 deadline);
+
+    /**
+     * @dev See {EIP712-constructor}.
+     */
+    constructor(string memory name) EIP712(name, "1") {}
 
-    function verify(ForwardRequest calldata req, bytes calldata signature) public view returns (bool) {
-        address signer = _recover(req, signature);
-        (bool correctFrom, bool correctNonce) = _validateReq(req, signer);
-        return correctFrom && correctNonce;
+    /**
+     * @dev Returns `true` if a request is valid for a provided `signature` at the current block.
+     */
+    function verify(ForwardRequest calldata request, bytes calldata signature) public view virtual returns (bool) {
+        (bool alive, bool signerMatch, bool nonceMatch) = _validate(request, signature);
+        return alive && signerMatch && nonceMatch;
     }
 
+    /**
+     * @dev Executes a `request` on behalf of `signature`'s signer guaranteeing that the forwarded call
+     * will receive the requested gas and no ETH is stuck in the contract.
+     */
     function execute(
-        ForwardRequest calldata req,
+        ForwardRequest calldata request,
         bytes calldata signature
-    ) public payable returns (bool, bytes memory) {
-        address signer = _recover(req, signature);
-        (bool correctFrom, bool correctNonce) = _validateReq(req, signer);
+    ) public payable virtual returns (bool, bytes memory) {
+        (bool success, bytes memory returndata) = _execute(request, signature);
 
-        if (!correctFrom) {
-            revert MinimalForwarderInvalidSigner(signer, req.from);
+        if (msg.value != request.value) {
+            revert MinimalForwarderMismatchedValue(request.value, msg.value);
```

Comment:
shouldn't we do this check before the `_execute()` call ?

Reply:
Yes, the ordering is result from multiple maneuvering I've been doing here. Changing.